src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -694,16 +695,24 @@ private synchronized AwsIotMqttClient getConnection(boolean forSubscription) {
             connections.add(conn);
             return conn;
         } else {
-            // Check for, and then close and remove any connection that has no subscriptions
-            Set<AwsIotMqttClient> closableConnections =
-                    connections.stream().filter((c) -> c.subscriptionCount() == 0).collect(Collectors.toSet());
-            for (AwsIotMqttClient closableConnection : closableConnections) {
-                // Leave the last connection alive to use for publishing
-                if (connections.size() == 1) {
-                    break;
+            // Check if there are more than 1 connections which can accept new subscriptions.
+            if (connections.stream().filter(AwsIotMqttClient::canAddNewSubscription).count() > 1) {
+                // Check for, and then close and remove any connection that has no subscriptions or any in progress
+                // subscriptions.
+                Set<AwsIotMqttClient> closableConnections =
+                        connections.stream()
+                                .filter((c) -> c.subscriptionCount() == 0 && c.inprogressSubscriptionsCount() == 0)
```

Comment:
isn't there a race here between updating and reading the subscriptionCount and inprogress count?